pkg/config.go

```diff
@@ -0,0 +1,83 @@
+package pkg
+
+import (
+	"fmt"
+	"github.com/spf13/viper"
+	"gopkg.in/validator.v2"
+)
+
+/*
+Config schema
+
+All values:
+- may be set in terra-status-manager.yaml, see cmd/root.go:
+  ```yaml
+  client:
+    redirects: 2
+  ```
+- may be overridden via environment variables:
+  ```bash
+  TSM_CLIENT_REDIRECTS=2
+  ```
+
+Some values:
+- may be overridden via command line flags, noted below.
+- may have non-"zero" default values, noted below and set in newDefaultConfig().
+- may be required to be non-"zero", noted below and validated in AssembleConfig().
+*/
+type Config struct {
+	// Whether to be more verbose with console output
+	// NOTE: May be set via --verbose / -v command line flags
+	Verbose bool
+
+	Client struct {
+		// Number of 300-series redirects to follow
+		Redirects int // default: 3
+		// Number of exponential-backoff retries to make
+		Retries int // default: 3
+	}
+
+	Statuspage struct {
+		// API key to communicate with Statuspage.io
+		// NOTE: Required but sensitive, should be set via TSM_STATUSPAGE_APIKEY=...
+		ApiKey string `validate:"nonzero"`
+		// ID of the particular page to interact with
+		PageID     string `validate:"nonzero"`
+		ApiRoot    string // default: "https://api.statuspage.io/v1"
+		Components []Component
+	}
+}
+
+// Component configuration--note that leaving any of the below unfilled will use Go's "zero" value (false/empty)
+type Component struct {
+	// Unique but user-readable component name
+	Name        string `validate:"nonzero"`
+	Description string
+	// If the component should be hidden to users while operational
+	OnlyShowIfDegraded bool
+	// If uptime data should be hidden and go unrecorded
+	HideUptime bool
+	// Date the component existed from, in the form YYYY-MM-DD
+	StartDate string `validate:"nonzero"`
+}
+
+func newDefaultConfig() *Config {
+	var c Config
+	c.Client.Redirects = 3
```

Comment:
I think it makes sense to stores these as `const` e.g. DEFAULT_REDIRECTS

Reply:
Really small hill and not one I'll really fight, just wanted to convey my thoughts. I actually disagree a tiny bit, for a few reasons:
- `const`s are exported from the package, and I don't think these defaults should be. From anywhere else, I think the only way to get a number of redirects should be via `config.Client.Redirects`--I know this isn't OO but I think the encapsulation could help avoid logical issues and keep the "namespace" clear.
- especially for non-IDE viewing, I think that `config.Client.Redirects = 3` is actually more useful than `const DEFAULT_REDIRECTS = 3` because it tells you how to change the value from a config file and how to access it from code. I know the const could be more verbose to convey the same info, but this way the type system helps us out and gives us one less layer of indirection to worry about maintaining.

Reply:
fair point on the `const`s, I was thinkings `const`s were just an immutable variable but it looks like there's some more complexity involved in using them.

I was imagining just having a couple of consts/statics up on the top of the file so we don't have orphaned numbers and then we can clutter the top of the file with comments and stuff rather than having it in the middle of logic. e.g. ```
// some more descriptive comment
const DEFAULT_REDIRECTS = 3
...some mount of code...
c.Client.Redirects = DEFAULT_REDIRECTS
```
so in case there was some reason you wanted to be wordy. e.g. `// we have to set redirects at least to 3 or above in order to not break world` you didn't have to do so in the middle of the func.

...er whats OO?

Reply:
I totally agree about documenting these values, especially when constraints are involved. I thought maybe we could collect this sort of info in one place, and I think the best place is where the variable is declared, where we can even do so in code:

```
type Config struct {
        // ...
	Client struct {
		// Number of 300-series redirects to follow, more explanation . . .
		Redirects int `validate:"gte=3"` // default: 3
```

I really wish that Go offered a way for me to have default values in a type declaration, but it doesn't. We can't have const structs either.

I think the most common use case for this file by far is people literally looking at it on github to understand how to make a `revere.yaml` file. To that end, I think we really need to justify redirection, not just in code but in how the documentation is read. 

We can't have a single point of declaration+validation+default (it's pick two, I think), but with the above at least we have a single point of documentation. Documenting the same field twice, or the default separate from the field, or some fields in the struct and some at a const, at the very least means that the user has to jump around the page, if not go and find the unexported function that wires up the consts and the structs to figure out how it works.

OO is "object oriented"--was just making a reference to encapsulation (but from playing around with this for a while it looks like we could have a lowercase unexported const).

I'm sorry for being pedantic about this--I think the experience of reading this specific code is one of the highest-impact controls on the success of this service. We can make APIs or pipe queue messages easily but this file being the backing for the user task of "write a YAML file that achieves compliance for communicating production outages" makes me put a really high (and wordy, I'm sorry) priority on getting it right.

Reply:
Pushing back is _perfectly okay_ as we're both working with the goal of making great readable/maintainable code/features, I'm happy to argue till the sun goes down because it helps flush out concerns or misconceptions. I totally agree with not double-documenting and _absolutely_ agree with the get-the-yaml-usage-correct statement. 

As I mentioned before I'm still picking up go so I don't know all the gotchas yet. Actually after reading the AssembleConfig comment, I think my understanding now is that `newDefaultConfig` _is_ the default config declaration, I think I got it into my head that newDefaultConfig was something other than an initialization method (too used to ruby i guess). Giving it more thought I think what you have is fine.